File: app/src/main/java/com/example/annada/eventfinder/adapters/CategoryAdapter.java
package com.example.annada.eventfinder.adapters;

import android.content.Context;
import android.support.v7.graphics.Palette;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.annada.eventfinder.R;
import com.example.annada.eventfinder.model.categoryapi.CategoryModel;
import com.example.annada.eventfinder.views.PaletteTransformation;
import com.squareup.picasso.Callback;
import com.squareup.picasso.NetworkPolicy;
import com.squareup.picasso.Picasso;

import java.util.List;

/**
 * Created by : annada
 * Date : 22/10/2017.
 */

public class CategoryAdapter extends RecyclerView.Adapter<CategoryViewHolders> {
    private List<CategoryModel> mCategories;
    private Context mContext;

    public CategoryAdapter(Context context, List<CategoryModel> categoryList) {
        mCategories = categoryList;
        mContext = context;
    }

    @Override
    public CategoryViewHolders onCreateViewHolder(ViewGroup parent, int viewType) {

        View rowView = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_view, parent, false);
        return new CategoryViewHolders(rowView);
    }

    @Override
    public int getItemCount() {
        return mCategories.size();
    }

    @Override
    public void onBindViewHolder(CategoryViewHolders holder, int position) {

        final CategoryModel selectedCategory = mCategories.get(position);
        final CategoryViewHolders finalHolder = holder;

        holder.mCategoryId = selectedCategory.getCategoryId();
        holder.categoryName.setText(selectedCategory.getCategoryName());
        Picasso.with(mContext)
                .load(selectedCategory.getCategoryIcon())
                .transform(PaletteTransformation.instance())
                //attempt to load the image from cache
                .networkPolicy(NetworkPolicy.OFFLINE)
                .fit()
                .into(finalHolder.categoryIcon, new PaletteTransformation.PaletteCallback(finalHolder.categoryIcon) {
                    @Override
                    public void onSuccess(Palette palette) {
                        int color = palette.getLightVibrantColor(
                                mContext.getResources().getColor(R.color.light_gray));
                        finalHolder.categoryIcon.setBackgroundColor(color);
                        selectedCategory.color = color;
                    }

                    @Override
                    public void onError() {
                        //Try again online if cache failed
                        Picasso.with(mContext)
                                .load(selectedCategory.getCategoryIcon())
                                .into(finalHolder.categoryIcon, new Callback() {
                                    @Override
                                    public void onSuccess() {
                                    }

                                    @Override
                                    public void onError() {
                                    }
                                });
                    }
                });

    }
}

File: app/src/main/java/com/example/annada/eventfinder/adapters/CategoryAdapter.kt
package com.example.annada.eventfinder.adapters

import android.content.Context
import android.support.v7.graphics.Palette
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.example.annada.eventfinder.R

import com.example.annada.eventfinder.model.categoryapi.CategoryModel
import com.example.annada.eventfinder.views.PaletteTransformation
import com.squareup.picasso.NetworkPolicy
import com.squareup.picasso.Picasso
import com.squareup.picasso.Callback
import kotlinx.android.synthetic.main.row_view.view.*


/**
 * Created by : annada
 * Date : 22/10/2017.
 */

class CategoryAdapter(private val mContext: Context, private val mCategories: List<CategoryModel>) : RecyclerView.Adapter<CategoryViewHolders>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryViewHolders {

        val rowView = LayoutInflater.from(parent.context).inflate(R.layout.row_view, parent, false)
        return CategoryViewHolders(rowView)
    }

    override fun getItemCount() = mCategories.size

    override fun onBindViewHolder(holder: CategoryViewHolders, position: Int) {

        val selectedCategory = mCategories[position]

        holder.mCategoryId = selectedCategory.categoryId
        holder.itemView.venue_name!!.text = selectedCategory.categoryName
        Picasso.with(mContext)
                .load(selectedCategory.categoryIcon)
                .transform(PaletteTransformation.instance())
                //attempt to load the image from cache
                .networkPolicy(NetworkPolicy.OFFLINE)
                .fit()
                .into(holder.itemView.venue_icon, object : PaletteTransformation.PaletteCallback(holder.itemView.venue_icon!!) {
                    public override fun onSuccess(palette: Palette) {
                        val color = palette.getLightVibrantColor(
                                mContext.resources.getColor(android.R.color.holo_orange_light))
                        holder.itemView.venue_icon!!.setBackgroundColor(color)
                        selectedCategory.color = color
                    }

                    override fun onError() {
                        //Try again online if cache failed
                        Picasso.with(mContext)
                                .load(selectedCategory.categoryIcon)
                                .into(holder.itemView.venue_icon, object : Callback {
                                    override fun onSuccess() {}

                                    override fun onError() {}
                                })
                    }
                })
    }
}
